zerver/lib/typed_endpoint.py

```diff
@@ -0,0 +1,514 @@
+    # type if there is one.
+    param_type: Type[T]
+    # The Pydantic TypeAdapter used to parse arbitrary input to the desired type.
+    # We store it on the FuncParam object as soon as the view function is
+    # decorated because it is expensive to construct.
+    # See also: https://docs.pydantic.dev/latest/usage/type_adapter/
+    type_adapter: TypeAdapter[T]
+
+    # The following group of attributes are computed from the ApiParamConfig
+    # annotation associated with this param:
+    # Name of the corresponding variable in the request data to look
+    # for. When argument_type_is_body is True, this is set to "request".
+    aliases: Tuple[str, ...]
+    argument_type_is_body: bool
+    documentation_status: DocumentationStatus
+    path_only: bool
+    request_var_name: str
+
+
+@dataclass(frozen=True)
+class ViewFuncInfo:
+    view_func_full_name: str
+    parameters: Sequence[FuncParam[object]]
+
+
+def is_annotated(type_annotation: Type[object]) -> bool:
+    origin = get_origin(type_annotation)
+    return origin is Annotated
+
+
+def parse_single_parameter(
+    param_name: str, param_type: Type[T], parameter: inspect.Parameter
+) -> FuncParam[T]:
+    param_default = parameter.default
+    # inspect._empty is the internal type used by inspect to indicate not
+    # specified defaults.
+    if param_default is inspect._empty:
+        param_default = NotSpecified
+
+    # Defaulting a value to None automatically wraps the type annotation with
+    # Optional. We explicitly unwrap it for the case of Annotated, which
+    # otherwise causes undesired behaviors that the annotated metadata gets
+    # lost. This is fixed in Python 3.11:
+    # https://github.com/python/cpython/issues/90353
+    if param_default is None:
+        origin = get_origin(param_type)
+        type_args = get_args(param_type)
+        if origin is Union and type(None) in type_args and len(type_args) == 2:
+            inner_type = type_args[0] if type_args[1] is type(None) else type_args[1]
+            if is_annotated(inner_type):
+                param_type = inner_type
+
+    param_config: Optional[ApiParamConfig] = None
+    if is_annotated(param_type):
+        # The first type is the underlying type of the parameter, the rest are
+        # metadata attached to Annotated. Note that we do not transform
+        # param_type to its underlying type because the Annotated metadata might
+        # still be needed by other parties like Pydantic.
+        _, *annotations = get_args(param_type)
+        for annotation in annotations:
+            if not isinstance(annotation, ApiParamConfig):
+                continue
+            assert param_config is None, "ApiParamConfig can only be defined once per parameter"
+            param_config = annotation
+    # Set param_config to a default early to avoid additional None-checks.
+    if param_config is None:
+        param_config = ApiParamConfig()
```

Comment:
How does this empty `ApiParamConfig` get used? Is there an example I should be thinking about?

Reply:
This is used for parameters that do not define `ApiParamConfig`. It applies to the majority of parameters typed as `str`, `Json[int]` and similar.

Reply:
OK, that makes sense -- probably worth a nice comment explaining this detail, and maybe also a mention of that detail in the `ApiParamConfig` definition that an empty one is used for parameters with no special configuration beyond their type.